lucille-core/src/main/java/com/kmwllc/lucille/stage/OpenAiEmbedding.java

```diff
@@ -0,0 +1,135 @@
+ */
+
+public class OpenAiEmbedding extends Stage {
+  // this is the token limit for all embedding models from openai
+  private static final int DEFAULT_OPENAI_TOKEN_LIMIT = 8191;
+  private final String API_KEY;
+  private EmbeddingModel model;
+  private final String textField;
+  private final boolean embedDocument;
+  private final boolean embedChildren;
+  private final String embeddingField;
+  private final OpenAIModels modelName;
+  private final Integer dimensions;
+  private Encoding enc;
+
+  private static final Logger log = LoggerFactory.getLogger(OpenAiEmbedding.class);
+
+  public OpenAiEmbedding(Config config) throws StageException {
+    super(config, new StageSpec().withRequiredProperties("text_field", "embed_document", "embed_children", "api_key"));
+    textField = config.getString("text_field");
+    embedDocument = config.getBoolean("embed_document");
+    embedChildren = config.getBoolean("embed_children");
+    API_KEY = config.getString("api_key");
+    embeddingField = config.hasPath("embedding_field") ? config.getString("embedding_field") : "embeddings";
+    modelName = OpenAIModels.fromConfig(config);
+    dimensions = config.hasPath("dimensions") ? config.getInt("dimensions") : null;
+    if (!embedDocument && !embedChildren) {
+      throw new StageException("Both embed_document and embed_children are false.");
+    }
+    if (API_KEY.isEmpty()) {
+      throw new StageException("API key is empty.");
+    }
+  }
+
+  @Override
+  public void start() throws StageException {
+    model = OpenAiEmbeddingModel.builder()
+        .modelName(modelName.getModelName())
+        .dimensions(dimensions)
+        .apiKey(API_KEY)
+        .build();
+
+    // retrieve encoding from modelName
+    ModelType modelType = ModelType.fromName(modelName.getModelName()).orElse(ModelType.TEXT_EMBEDDING_3_SMALL);
+    enc = Encodings.newDefaultEncodingRegistry().getEncodingForModel(modelType);
+  }
+
+  @Override
+  public Iterator<Document> processDocument(Document doc) throws StageException {
+    if (embedDocument) {
+      sendForEmbedding(doc);
+    }
+    if (embedChildren && doc.hasChildren()){
+      List<Document> childDocs = doc.getChildren();
+      for (Document childDoc : childDocs) {
+        sendForEmbedding(childDoc);
+      }
+    }
+    return null;
+  }
+
+  private void sendForEmbedding(Document doc) throws StageException {
+    if (!doc.has(textField)) {
+      log.warn("{} does not exist, skipping document: {}", textField, doc);
+      return;
+    }
+
+    String content = doc.getString(textField);
+    content = ensureContentIsWithinLimit(content);
+    Response<Embedding> response;
+
+    // model.embed will retry 3 times by default
+    // model.embed might throw runtimeException due to invalid API key and other errors.
+    try {
+      response = model.embed(content);
+    } catch (RuntimeException e) {
+      log.warn("failed to get embedding", e);
```

Comment:
First, this log message should be improved to be more informative. maybe include the id of the erroring document?

Second, I think we should throw a StageException here instead of just logging. This error would mean that none of our encodings are being properly created, so we would likely not want to just skip this step and continue the run if this is not working.